pkg/standalone-utils/contracts/PoolRecoveryEnabler.sol

```diff
@@ -0,0 +1,94 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IBasePoolFactory.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IRateProviderPool.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IRecoveryMode.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/SingletonAuthentication.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/EnumerableSet.sol";
+
+/**
+ * @dev This contract allows anyone to set Pools that feature rate providers into recovery mode, provided any of said
+ * rate providers revert. This helps identify failing Pools to both let LPs promptly exit them, and also prevent further
+ * swaps from being routed through them by off-chain mechanisms (since said swaps would fail).
+ */
+contract PoolRecoveryEnabler is SingletonAuthentication {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    EnumerableSet.AddressSet private _factories;
+
+    constructor(IVault vault, address[] memory initialFactories) SingletonAuthentication(vault) {
+        for (uint256 i = 0; i < initialFactories.length; ++i) {
+            require(_factories.add(initialFactories[i]), "Duplicate initial factory");
+        }
+    }
+
+    function addPoolFactory(address factory) external authenticate {
+        require(_factories.add(factory), "Duplicate factory");
+    }
+
+    function removePoolFactory(address factory) external authenticate {
+        require(_factories.remove(factory), "Non-existent factory");
+    }
+
+    function getFactoryAt(uint256 index) external view returns (IBasePoolFactory) {
+        return IBasePoolFactory(_factories.at(index));
+    }
+
+    function getFactoryCount() external view returns (uint256) {
+        return _factories.length();
+    }
+
+    function isPoolFromKnownFactory(address pool) public view returns (bool) {
+        uint256 totalFactories = _factories.length();
+        for (uint256 i = 0; i < totalFactories; ++i) {
+            IBasePoolFactory factory = IBasePoolFactory(_factories.unchecked_at(i));
+
+            if (factory.isPoolFromFactory(pool)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    function enableRecoveryModeInPool(address pool) external {
```

Comment:
Yes, didn't want to do this until the name was finalized, but thanks for pointing it out.

Not fond of `verify` since it doesn't verify anything: it puts a Pool in recovery mode (or reverts).

Reply:
Well, it verifies that the rate providers work (or don't) - but yes, it's confusable with other senses of verify. I had `checkRateProviders` first, but that doesn't mention the crucial feature of putting it in recovery mode.

We could just spell it out: `enableRecoveryModeOnPoolWithFailingRateProvider`

What if a rate provider had a temporary problem (or was upgraded) and fixed itself? Governance would have to disable Recovery Mode. We *could* have a parallel function called `disableRecoveryMode...` with reversed logic, that would revert immediately if not in Recovery Mode, and otherwise check all the rate providers, disabling Recovery Mode if they were all good, and reverting if any rate providers reverted.

Reply:
I wouldn't mind having governance intervene in such a situation to address the issue. I don't like giving too much power over these things to permissionless contracts - that could potentially lead to DoS attacks etc.

The Pool being in recovery mode is a non-issue -- all that happens is protocol fees are forfeit. If a Pool had a critical issue that required it being put into recovery mode, I think it's fine for governance to step in and review what happened before regular operation is restored.